core-common/src/main/java/org/glassfish/jersey/message/internal/AbstractMessageReaderWriterProvider.java

```diff
@@ -42,18 +43,24 @@
 
     /**
      * The UTF-8 Charset.
+     *
+     * @deprecated use {@code StandardCharsets.UTF8} instead.
      */
-    public static final Charset UTF8 = ReaderWriter.UTF8;
+    @Deprecated(forRemoval = true)
+    public static final Charset UTF8 = StandardCharsets.UTF_8;
 
     /**
      * Reader bytes from an input stream and write then to an output stream.
      *
      * @param in  the input stream to read from.
      * @param out the output stream to write to.
      * @throws IOException if there is an error reading or writing bytes.
+     *
+     * @deprecated use {@code in.transferTo(out)} instead.
      */
+    @Deprecated
     public static void writeTo(InputStream in, OutputStream out) throws IOException {
-        ReaderWriter.writeTo(in, out);
+        in.transferTo(out);
```

Comment:
In general, I am not in favor of replacing `ReaderWriter.writeTo`. While transferTo seems to do exactly the same as `writeTo`, Jersey has a mechanism to extend the buffer size used.

Reply:
In a JMH test campaign at OpenJDK we have empirically *proven* that it makes *no sense* for any application to use any other buffer sizes than the JRE's default size. In particular we have proven that it makes *no sense* to use buffers smaller than 8K (JRE's previous default) / larger than 16K (JRE's current default). Keeping that configuration option will *in virtually all* cases result in (very) worse efficience and performance compared to using `transferTo`. Hence I do not see what your target is by *not* using `transferTo`. Using `transferTo` is the *sole* key point of this PR, and is the *sole proven* driver for optimum efficience and performance. So if you decision is final, we can drop this PR -- but this means that Jersey will always have bad efficency and performance.

Reply:
BTW, transferTo is definitively *not* doing the same as Jersey's custom code in recent JREs. In fact, modern JREs offload the data transfer to the OS, which leads to dramatic performance and efficiencecy gains. My PR targets on moden JREs, not on JRE 11. That's why I opened it for 3.1, not for 3.x / master. Please check the actual source code of JRE 21-SNAPSHOT.

Reply:
A compromise could be that we keep calling `ReaderWriter.transferTo` but inside of that method we add an option that is calling the old code *if explicitly wanted*. For example, we could replace the default buffer size 8192 by -1. For -1 we call transferTo. For any other number (hence explicitly set 8192 or any other value) we call the original code. WDYT?

Reply:
@jansupol I think this is what we could agree upon: c118b65a235d9a2d4ef9ada743312998a2ce04b0. It preserves the option to explicitly set the buffer size, but unless one does so, it let's the JRE do its job. So user who don't care get JRE optimizations out-of-the-box, and users who want to get the old code, can explicity set 8192. This is the best of all worlds.

Reply:
NB: User should abstain from actually setting *particularly* the original value of 8192. If one really has tested different buffer sizes carefully (like we did at OpenJDK), he will notice that **16384** *always* outperfoms 8192, while bigger buffers do not provide *considerable* additional gain (but cost a lot of RAM in sum on production servers with massive parallel requests). Due to that, modern JREs use 16384 (at least for byte buffers at time of writing).

Reply:
@mkarg ReaderWriter's BUFFER_SIZE is a public property with a default value of 8192. This must not be changed, think about compatibility.

I do not see much difference in my early copy of JDK 21. But I agree, the JDK internals can change. This is seen by extending the buffer values in JDK 21. While the tests with JDK may show the large buffer brings a slight performance boost, the customer may judge otherwise, based on their use case. 

You are right, however, that by default the buffer size is not changed by the majority of customers. Hence, what we can do is forward to JDK's transfer method for cases where the customer does not change the BUFFER_SIZE to a value different from 8192.

Reply:
@jansupol I see we come nearer to a final solution. :-)

What we measured at OpenJDK is that the performance / efficiency benefit strongly depends greatly on the actual *platform stack*, less of the actual *use case*, which is why we still have decided for a *one size fits it all* buffer size even in JDK 21. I do not see any evidence that this experience should not fit to particularly Jersey. Anyways I have no objection against keeping the IO_BUFFER_SIZE option available for those rare cases that I might not see but possibly exist (do you actually know some?).

The actual code changes BTW can be found here: https://github.com/openjdk/jdk/pulls?q=is%3Apr+is%3Aclosed+author%3Amkarg. They (mostly) are effective *solely* when using transferTo, in particular when using transferTo in conjunction with JRE (non-custom) streams (hence the lots of changed implementations) - as you can see, over the past years, we changed and discussed a lot of JRE code, and that for a good reason (to make the most of modern server hardware / to not stand in the way of OS-internal optimizations). 

Regarding the default size: I do not think that anybody really relies on particularly the value 8192 being the default for all times, but people instead rely on "BUFFER_SIZE == IO_DEFAULT_BUFFER_SIZE", and they rely on IO_DEFAULT_BUFFER_SIZE telling them the *actual* value at any time they can trust upon *to be the default*, so I think it would be OK if we tell them that IO_DEFAULT_BUFFER_SIZE == -1 from the next minor or major release. IMHO existing applications that did *not* check IO_DEFAULT_BUFFER_SIZE are simply *wrong* because that is the reason why the docs do *not* say "8192" but say "IO_DEFAULT_BUFFER_SIZE", right?

Only enabling the fallback code in case the application is setting a value *other than 8192* sounds rather odd and ambiguous. What if a user *wants* to really use just 8192 bytes, because he does not want to squander the new JRE default of 16384 (or possibly even more bytes in JRE 22 or 23)? So I really would recommend to go with IO_DEFAULT_BUFFER_SIZE == -1 instead, to make the actual behavior change *visible* (if we trust our users to read the release notes), or (for maximum backwards compatibility) to *actually detect **any** change* by using a Jersey-internal flag which is set to true as soon as IO_BUFFER_SIZE >= 0 (i. e. Jersey *internally* ignores BUFFER_SIZE, but externally the BUFFER_SIZE backward-compatibly reflects IO_DEFAULT_BUFFER_SIZE == 8192).

**Having said that, in what direction shall I effectively migrate this PR now?**

Reply:
@jansupol "ping" - I need your decision regarding above proposal. Thanks.

Reply:
Really it is not a good practice to remove or change public properties, and values. 

You have a point that people might require a lower buffer than the one in JDK on purpose. However, since the change comes in JDK 21 which is not released yet, it is not very likely people will want to override that value, and it makes more sense to me to live with overriding this value than changing public constants. 

If you want to be precise, you can check for JDK version (using the Jersey JdkVersion class) and if JDK is 21, keep Jersey routine for BUFFER_SIZE 16384. But then, is it not the goal to use JDK code instead of Jersey? While I can imagine the poor guy spending days on testing various BUFFER_SIZE with JDK 11, I can hardly imagine that guy doing so with beta JDK 21.